docs/Style.thy

```diff
@@ -0,0 +1,260 @@
+   my_function param_a param_b
+   \<lbrace> post_condition \<rbrace>"
+  oops
+
+text \<open> or \<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace> precondition_one and precondition_two
+     and precondition three \<rbrace>
+   my_function param_a param_b
+   \<lbrace> post_condition \<rbrace>"
+  oops
+
+text \<open>
+Definitions, record and datatypes are generally indented as follows and use \<equiv> rather than =. Short
+datatypes can be on a single line.
+\<close>
+
+definition word_bits :: nat where
+  "word_bits \<equiv> 32"
+
+definition long_defn_type ::
+  "'this => 'is => 'a => 'long => 'type => 'that => 'does =>
+   'not => 'fit => 'in => 'one => 'line"
+  where
+  "long_defn_type \<equiv> undefined"
+
+fun foo_fun :: "nat \<Rightarrow> nat" where
+  "foo_fun 0 = 0" |
+  "foo_fun (Suc n) = n"
+
+record cdl_cnode =
+  cdl_cnode_caps :: type_foo
+  cdl_cnode_size_bits :: type_bar
+
+datatype cdl_object =
+    Endpoint
+  | Tcb type_foo
+  | CNode cdl_cnode
+  | Untyped
+
+datatype cdl_arch = IA32 | ARM11
+
+text \<open>
+There are tools to automatically indent proofs in both jEdit and Emacs. In terms of rules:
+  * definition and fun should be on the same line as the name.
+  * Everything after the first line of a definition or fun statement should be indented.
+  * Type definitions (record, type_synonym, datatype) have the defined type on the same line, the
+    rest only if it fits.
+  * Avoid long hanging indentation in data types, case expressions, etc. That is:
+
+Wrong:
+
+@{term\<open>
+case xs_blah of [] \<Rightarrow> foo
+              | y#ys \<Rightarrow> baz
+\<close>}
+
+Right:
+
+@{term\<open>
+case xs_blah of
+    [] \<Rightarrow> foo
+  | y#ys \<Rightarrow> baz
+\<close>}
+\<close>
+
+text \<open>
+  If a definition body causes an overflow of the 100 char line limit, use one
+  of the following two patterns.
+\<close>
+
+definition hd_opt :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt \<equiv> case_list
+            None
+            (\<lambda>h t. Some h)"
+
+definition hd_opt4 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt4 \<equiv>
+   case_list None (\<lambda>h t. Some h)"
+
+text \<open>
+  If an apply line causes an overflow of the 100 char line limit, use one
+  of the following two patterns: aligning on ":" or aligning on left.
+\<close>
+
+lemma test_lemma0:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+    hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                 split: list.splits)
+  done
+
+lemma test_lemma1:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+    hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                  split: list.splits)
+  done
+
+lemma test_lemma3:
+  "case_option None hd_opt (hd_opt l) = Some x \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply ((case_tac l; simp add: hd_opt_def),
+         rename_tac h t,
+         case_tac h; simp)
```

Comment:
Do we want to require extra bracketing when `;` and `,` interact? I don't know about others, but I haven't managed to internalise this yet and it's not obvious to me whether these lines are correctly indented.

Reply:
I'm actually not that sure either. It'd be better to figure that out and document it better than requiring more parentheses, though.

Reply:
These parentheses are required since `,` apparently binds stronger than `;`. How could I improve this example, or should I remove the example?